pkg/util/scheduler/scheduler.go

```diff
@@ -0,0 +1,203 @@
+	subservicesWatcher *services.FailureWatcher
+
+	workers    []*Worker
+	numWorkers int
+}
+
+// Config holds configuration for the Scheduler.
+type Config struct {
+	NumWorkers int
+	MaxBackoff time.Duration
+	MaxRetries int
+	Logger     log.Logger
+}
+
+func (c *Config) validate() error {
+	if c.NumWorkers <= 0 {
+		c.NumWorkers = DefaultNumWorkers
+	}
+	if c.MaxBackoff <= 0 {
+		c.MaxBackoff = DefaultMaxBackoff
+	}
+	if c.MaxRetries <= 0 {
+		c.MaxRetries = DefaultMaxRetries
+	}
+	if c.Logger == nil {
+		c.Logger = log.New("scheduler")
+	}
+	return nil
+}
+
+// NewScheduler creates a new scheduler instance.
+func NewScheduler(queue WorkQueue, config *Config) (*Scheduler, error) {
+	var err error
+
+	if queue == nil {
+		return nil, errors.New("queue cannot be nil")
+	}
+	if err := config.validate(); err != nil {
+		return nil, fmt.Errorf("invalid config: %w", err)
+	}
+
+	s := &Scheduler{
+		logger:             config.Logger,
+		queue:              queue,
+		numWorkers:         config.NumWorkers,
+		maxBackoff:         config.MaxBackoff,
+		workers:            make([]*Worker, 0, config.NumWorkers),
+		subservicesWatcher: services.NewFailureWatcher(),
+	}
+
+	subservices := []services.Service{s.queue}
+	s.subservices, err = services.NewManager(subservices...)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create subservices manager: %w", err)
+	}
+	s.Service = services.NewBasicService(s.starting, s.running, s.stopping)
+	return s, nil
+}
+
+// starting is called by the services.Service lifecycle to start the scheduler.
+func (s *Scheduler) starting(ctx context.Context) error {
+	s.subservicesWatcher.WatchManager(s.subservices)
+
+	if err := services.StartManagerAndAwaitHealthy(ctx, s.subservices); err != nil {
+		return fmt.Errorf("scheduler: failed to start subservices: %w", err)
+	}
+
+	s.logger.Info("scheduler starting", "numWorkers", s.numWorkers)
+	s.workers = make([]*Worker, 0, s.numWorkers)
+	s.wg.Add(s.numWorkers)
+
+	for i := 0; i < s.numWorkers; i++ {
+		worker := &Worker{
+			id:         i,
+			queue:      s.queue,
+			wg:         &s.wg,
+			maxBackoff: s.maxBackoff,
+			logger:     s.logger,
+		}
+		s.workers = append(s.workers, worker)
+		go worker.run(ctx)
+	}
+
+	s.logger.Info("scheduler started")
+	return nil
+}
+
+// running is called by the services.Service lifecycle to check if the scheduler is running.
+func (s *Scheduler) running(ctx context.Context) error {
+	select {
+	case <-ctx.Done():
+		return nil
+	case err := <-s.subservicesWatcher.Chan():
+		return fmt.Errorf("subservice failure: %w", err)
+	}
+}
+
+// stopping is called by the services.Service lifecycle to stop the scheduler.
+func (s *Scheduler) stopping(_ error) error {
+	s.logger.Info("scheduler stopping")
+
+	err := services.StopManagerAndAwaitStopped(context.Background(), s.subservices)
+	if err != nil {
+		return fmt.Errorf("scheduler: failed to stop subservices: %w", err)
+	}
+
+	s.logger.Info("scheduler stopped")
```

Comment:
Shall we wait until all workers are finished? (They should already stop soon, since they work with services' context, but maybe they run some runnable that's taking a long time.)